src/gcodeExport.cpp

```diff
@@ -1434,44 +1434,73 @@ void GCodeExport::writeFanCommand(double speed, std::optional<size_t> extruder)
-
-    if (iterator != current_fans_speeds_.end() && std::abs(iterator->second - speed) < 0.1)
-    {
-        return;
-    }
+    const auto iterator = current_fans_speeds_.find(fan_number);
+    const std::optional<double> current_fan_speed = (iterator != current_fans_speeds_.end()) ? std::optional<double>(iterator->second) : std::nullopt;
 
     if (flavor_ == EGCodeFlavor::MAKERBOT)
     {
-        if (speed >= 50)
+        // Makerbot cannot PWM the fan speed, only turn it on or off
+
+        bool write_value = true;
+        const bool new_on = speed >= 50;
+        if (current_fan_speed.has_value())
         {
-            *output_stream_ << "M126 T0" << new_line_; // Makerbot cannot PWM the fan speed...
+            const bool old_on = current_fan_speed.value() >= 50;
+            write_value = new_on != old_on;
         }
-        else
+
+        if (write_value)
         {
-            *output_stream_ << "M127 T0" << new_line_;
+            if (new_on)
+            {
+                *output_stream_ << "M126 T0" << new_line_;
+            }
+            else
+            {
+                *output_stream_ << "M127 T0" << new_line_;
+            }
         }
     }
-    else if (speed > 0)
+    else
     {
         const bool should_scale_zero_to_one = Application::getInstance().current_slice_->scene.settings.get<bool>("machine_scale_fan_speed_zero_to_one");
-        *output_stream_ << "M106 S"
-                        << PrecisionedDouble{ (should_scale_zero_to_one ? static_cast<uint8_t>(2) : static_cast<uint8_t>(1)),
-                                              (should_scale_zero_to_one ? speed : speed * 255) / 100 };
-        if (fan_number)
+        bool write_value = true;
+        std::ostringstream new_value;
+        new_value << PrecisionedDouble{ (should_scale_zero_to_one ? static_cast<uint8_t>(2) : static_cast<uint8_t>(1)), (should_scale_zero_to_one ? speed : speed * 255) / 100 };
+        const std::string new_value_str = new_value.str();
+        if (current_fan_speed.has_value())
         {
-            *output_stream_ << " P" << fan_number;
+            std::ostringstream old_value;
+            old_value << PrecisionedDouble{ (should_scale_zero_to_one ? static_cast<uint8_t>(2) : static_cast<uint8_t>(1)),
+                                            (should_scale_zero_to_one ? current_fan_speed.value() : current_fan_speed.value() * 255) / 100 };
+
+            write_value = new_value_str != old_value.str();
         }
-        *output_stream_ << new_line_;
-    }
-    else
-    {
-        *output_stream_ << "M107";
-        if (fan_number)
+
+        if (write_value)
         {
-            *output_stream_ << " P" << fan_number;
+            // Check if the value to be written is only made with zeroes, in which case it is actually a turn off
+            std::string new_value_str_reduced = new_value_str;
+            std::erase(new_value_str_reduced, '0');
+            std::erase(new_value_str_reduced, '.');
```

Comment:
Depending on the printer, values can range from 0.0-1.0 or 0-255 -- I think this would 'reduce' something like `205` to `25`, which isn't what you'd want.

Also not a fan (pun not originally intended) of the way we handle something numeric via string-manipulation here.

Reply:
Also will reduce `0.5` to just `5`.

Reply:
Well, that's precisely the point. I think the variable is not properly named. The goal of this is to identify anything that would turn into a numeric "0" value, would it be "0", "0.0" or "0.00". If the numeric value is e.g. 0.02, turning it to a one-digit precision string would be "0.0" so in this case I want to issue a M107 command and not M106.

I'm also not a big fan of doing it via string manipulation, but this is a way to ensure that you will never ever end up with a "M106 0.0".